### PR TITLE
zoekt-mirror-gerrit: handle http authentication

### PR DESCRIPTION
**Pull request by xavier-calland**
issue: #763

### CODE REVIEW

File: cmd/zoekt-mirror-gerrit/main.go [hunk 1]
```diff
@@ -224,3 +235,41 @@ func marshalBool(b bool) string {
...
+	}
+}
+
+func anonymousURL(URL *url.URL) string {
+	anonymousURL := *URL
+	if _, has := URL.User.Password(); has {
+		anonymousURL.User = nil
+	}
+	return (&anonymousURL).String()
+}
```
Comment (keegancsmith) [inline cmd/zoekt-mirror-gerrit/main.go]: what if just user is set? Wouldn't you want to strip it off then? If that is the case this is a simpler implementation:
```suggestion
func anonymousURL(u *url.URL) string {
  anon := *u
  anon.User = nil
  return anon.String()
}
```

**Review by keegancsmith — CHANGES_REQUESTED** [review]
It doesn't look like you need the credential helper that writes to disk. You can instead generate authenticated clone URLs. This would be a simpler implementation if I am not mistaken?

**Comment by xavier-calland** [comment]
> It doesn't look like you need the credential helper that writes to disk. You can instead generate authenticated clone URLs. This would be a simpler implementation if I am not mistaken?

Yes it works too.

But I'm not a fan of putting the password in plain text in the address
There is a risk of secrets leaking to the newspapers.
Additionally, the code would be closely tied to the git implementation.
Here we could more easily integrate with other git credential helpers.

**Comment by keegancsmith** [comment]
They both will communicate the credentials over the wire in the same way. The only difference I can tell is we store the credentials as part of the clone URL  in ./.git/config vs ./credential-store. FYI when I suggest using the clone URL then one does have to clean up the code path which extracts repo name from cloneURL to not include secrets.

**Comment by xavier-calland** [comment]
> FYI when I suggest using the clone URL then one does have to clean up the code path which extracts repo name from cloneURL to not include secrets.

I'm not sure I understand what you mean.

What is expected for the config values:

* remote.url (clone URL): `https://user:password@gerrit-host/repo`
* zoekt.gerrit-host: `https://gerrit-host/repo`

**Comment by keegancsmith** [comment]
> What is expected for the config values

Yes what you listed. Sorry I need to read the gerrit code much closer, but I took a cursory look at noticed the cloneURL was directly used to construct the name/etc. So without checking further I was just mentioning some care may need to be taken to avoid leaking auth into the name of the repo/etc.

**Comment by xavier-calland** [comment]
OK, done.

File: cmd/zoekt-mirror-gerrit/main.go [hunk 2]
```diff
@@ -224,3 +229,15 @@ func marshalBool(b bool) string {
...
+	anon := *u
+	anon.User = nil
+	return anon.String()
+}
+
+func addPassword(u string, user *url.Userinfo) string {
+	password, _ := user.Password()
+	username := user.Username()
+	return strings.Replace(u, fmt.Sprintf("://%s@", username), fmt.Sprintf("://%s:%s@", username, password), 1)
+}
```
Comment (keegancsmith) [inline cmd/zoekt-mirror-gerrit/main.go:243]: This is prone to breaking, I suggest using url.URL
```suggestion
func addPassword(rawURL string, user *url.Userinfo) string {
    u, err := url.Parse(rawURL)
    if err != nil {
        return rawURL
    }
    u.User = user
    return u.String()
}
```

**Review by keegancsmith — APPROVED** [review]

File: cmd/zoekt-mirror-gerrit/main.go (hunk 2 above)
Comment (xavier-calland) [reply cmd/zoekt-mirror-gerrit/main.go:243]: The problem with this solution is that il will encode the path and `${project}` is transformed to `$%7Bproject%7D` and break all `strings.Replace(…, "${project}")`

I can replace `$%7Bproject%7D`  → `${project}` before `return` (but I'm not sure it's a good solution)
or I can use `$%7Bproject%7D`  everywhere in the file
(I'm not convinced by either solution)

(besides, I noticed that the management of paths (host + repo name) could surely be simplified to centralize part of the logic)

What do you think is the solution?

Comment (keegancsmith) [reply cmd/zoekt-mirror-gerrit/main.go:243]: Alright I see now why we avoided URL parsing before, its not really a URL but a URL template. I think we will stick with this then, but a potential better solution is to only add in the password at the time we create the cloneURL.

**Comment by keegancsmith** [comment]
FYI I don't have any experience with the gerrit APIs (only used it as a normal user), hence why my reviews are not that great on this :) If you wanna keep making improvements to this go ahead :) Maybe a few simple tests can help with making bigger changes more confidently.

**Comment by xavier-calland** [comment]
I just remembered another reason why I didn't choose the option to put the password in the repository URL.
With this solution the password will not be updated if it changes.

**Comment by xavier-calland** [comment]
To make sure the credentials are ok we can:

1. force `remote.origin.url` in https://github.com/sourcegraph/zoekt/blob/72f95004e6d6136fed7bf973616e89e6d37e4eaa/gitindex/clone.go#L63-L66
2. force `remote.origin.url` in https://github.com/sourcegraph/zoekt/blob/72f95004e6d6136fed7bf973616e89e6d37e4eaa/cmd/zoekt-mirror-gerrit/main.go#L214-L218
3. use git credential

@keegancsmith What solution should I make a PR for?

**Comment by keegancsmith** [comment]
Making it update origin in the case of calling clone on an existing repo makes sense to me.

**Pull request merged**